receiver/awscontainerinsightreceiver/internal/k8sapiserver/k8sapiserver.go

```diff
@@ -16,54 +16,297 @@ package k8sapiserver
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/k8s/k8sclient"
+)
+
+const (
+	lockName = "otel-container-insight-clusterleader"
 )
 
+// eventBroadcaster is adpated from record.EventBroadcaster
+type eventBroadcaster interface {
+	// StartRecordingToSink starts sending events received from this EventBroadcaster to the given
+	// sink. The return value can be ignored or used to stop recording, if desired.
+	StartRecordingToSink(sink record.EventSink) watch.Interface
+	// StartLogging starts sending events received from this EventBroadcaster to the given logging
+	// function. The return value can be ignored or used to stop recording, if desired.
+	StartLogging(logf func(format string, args ...interface{})) watch.Interface
+	// NewRecorder returns an EventRecorder that can be used to send events to this EventBroadcaster
+	// with the event source set to the given event source.
+	NewRecorder(scheme *runtime.Scheme, source v1.EventSource) record.EventRecorder
+}
+
 // K8sAPIServer is a struct that produces metrics from kubernetes api server
 type K8sAPIServer struct {
+	nodeName            string //get the value from downward API
 	logger              *zap.Logger
 	clusterNameProvider clusterNameProvider
 	cancel              context.CancelFunc
+
+	mu      sync.Mutex
+	leading bool
+
+	k8sClient *k8sclient.K8sClient
+
+	// the following can be set to mocks in testing
+	broadcaster eventBroadcaster
+	// the close of isLeadingC indicates the leader election is done. This is used in testing
+	isLeadingC chan bool
 }
 
 type clusterNameProvider interface {
 	GetClusterName() string
 }
 
+type k8sAPIServerOption func(*K8sAPIServer)
+
 // New creates a k8sApiServer which can generate cluster-level metrics
-func New(clusterNameProvider clusterNameProvider, logger *zap.Logger) *K8sAPIServer {
+func New(clusterNameProvider clusterNameProvider, logger *zap.Logger, options ...k8sAPIServerOption) (*K8sAPIServer, error) {
 	_, cancel := context.WithCancel(context.Background())
 	k := &K8sAPIServer{
 		logger:              logger,
 		clusterNameProvider: clusterNameProvider,
+		k8sClient:           k8sclient.Get(logger),
+		broadcaster:         record.NewBroadcaster(),
 		cancel:              cancel,
 	}
 
-	if err := k.start(); err != nil {
-		k.logger.Warn("Fail to start k8sapiserver", zap.Error(err))
-		return nil
+	for _, opt := range options {
+		opt(k)
 	}
 
-	return k
+	if k.k8sClient == nil {
+		return nil, errors.New("failed to start k8sapiserver because k8sclient is nil")
+	}
+
+	if err := k.init(); err != nil {
+		return nil, fmt.Errorf("fail to initialize k8sapiserver, err: %v", err)
+	}
+
+	return k, nil
 }
 
 // GetMetrics returns an array of metrics
 func (k *K8sAPIServer) GetMetrics() []pdata.Metrics {
-	// TODO: add the logic to generate the metrics
 	var result []pdata.Metrics
+
+	// don't generate any metrics if the current collector is not the leader
+	if !k.leading {
```

Comment:
```suggestion
	k.mu.Lock()
	leading := k.leading
	k.mu.Unlock()
	if !leading {
```

Access to `k.leading` needs to be guarded with the mutex here.

Reply:
Right! I made the changes accordingly.